// lixi-core-nodejs/module/hub/instance/register_rest_rsa.js
/** đăng ký điểm giao tiếp */

let Joi = require('../../joi/_index');
let Axios = require('axios');

// #region [schema]

let schema_option = Joi.object({
    key: Joi.string()
        .required()
        .description('mã định danh'),
    host: Joi.string()
        .required()
        .description('host sẽ request'),
    path: Joi.string()
        .regex(/^\//, '"must starts with /"')
        .required()
        .description('path sẽ request'),
    method: Joi.string()
        .valid('GET', 'POST', 'PUT', 'DELETE')
        .required()
        .description('http method sử dụng'),
    data: Joi.any()
        .default({})
        .description('data mặc định'),
    header: Joi.object()
        .pattern(Joi.string(), Joi.string())
        .description('http header mặc định')
        .default({}),
    response_data_field: Joi.string()
        .default('data')
        .description('field chứa dữ liệu trong response'),
    response_error_field: Joi.string()
        .default('error')
        .description('field chứa lỗi trong response'),
    public_key_handler: Joi.func()
        .arity(1)
        .required()
        .description('hàm lấy public key để mã hóa gói tin'),
    private_key_handler: Joi.func()
        .arity(1)
        .required()
        .description('hàm lấy private key để mã hóa gói tin'),
    signature_handler: Joi.func()
        .arity(1)
        .description('hàm xử lý cấu trúc chữ ký')
});
schema_option = schema_option.default(schema_option.validate({}).value);

// #endregion

module.exports = hub =>
    (hub.registerRestRsa = function(options) {
        options = Joi.validateValue({ options }, schema_option);

        // #region [kiểm tra key tồn tại]

        if (hub.endpoints.some(x => x.key === options.key))
            throw Error(`hub endpoint "${options.key}" is already existed`);

        // #endregion

        // #region [khởi tạo endpoint]

        let endpoint = {
            key: options.key,
            protocol: 'rest_rsa',
            host: options.host,
            path: options.path,
            method: options.method,
            header: options.header,
            data: options.data,
            response_data_field: options.response_data_field,
            response_error_field: options.response_error_field,
            public_key_handler: options.public_key_handler,
            private_key_handler: options.private_key_handler,
            signature_handler: options.signature_handler
        };

        endpoint.client = Axios.create({
            baseURL: options.host + options.path,
            headers: options.header,
            data: options.data
        });

        // #endregion

        hub.endpoints.push(endpoint);
    });
