// lixi-core-nodejs/module/document_server/instance/register_module.js
/** ghi nhận thông tin về module */

let Joi = require('../../joi/_index');
let Callsite = require('callsite');
let Path = require('path');

// #region [schema]

let schema_option = Joi.object({
    dir: Joi.string().description('đường dẫn đến thư mục gốc của module'),
    description: Joi.string().description('mô tả module'),
    version: Joi.string().description('phiên bản module'),
    author: Joi.string().description('tác giả module'),
    list_dirname_biz: Joi.array()
        .items(Joi.string())
        .default(['biz', 'core'])
        .description('đường dẫn thư mục biz'),
    hub: Joi.object().description('hub của module'),
    apis: Joi.array()
        .items(Joi.object())
        .description('danh sách các api của module'),
    models: Joi.array()
        .items(Joi.object())
        .description('danh sách các model của module')
});
schema_option = schema_option.default(schema_option.validate({}).value);

// #endregion

module.exports = document =>
    (document.registerModule = function(option) {
        option = Joi.validateValue({ option }, schema_option);

        // #region [xác định tên module]

        if (!option.dir) option.dir = Path.dirname(Callsite()[1].getFileName());

        let module_name = option.dir
            .substring(document.info.dir_module.length + 1)
            .split('/')[0];

        // #endregion

        // #region [find or insert]

        let modu = document.info.modules.find(x => x.name === module_name);
        if (!modu) {
            modu = { name: module_name };
            document.info.modules.push(modu);
        }

        // #endregion

        // #region [set basic infomation]

        modu.dir = option.dir;
        modu.description = option.description;
        modu.version = option.version;
        modu.author = option.author;
        modu.bizs = option.list_dirname_biz || [];
        modu.apis = option.apis || [];
        modu.models = option.models || [];
        modu.hub_endpoints = option.hub ? option.hub.endpoints : [];

        // #endregion
    });
