// lixi-core-nodejs/module/api_server/instance/start.js
/** khởi chạy server */

let RequestLog = require('request-log');

module.exports = server =>
    (server.start = async function() {
        // #region [cài đặt request log]

        if (
            server.options.request_log_host &&
            server.options.request_log_table_name
        ) {
            try {
                server.request_log = new RequestLog.Custom();

                await server.request_log.setup({
                    enable: server.options.request_log_enable,
                    canLog: false,
                    serviceName: server.options.request_log_table_name,
                    connection: {
                        host: server.options.request_log_host,
                        keepAlive: false
                    }
                });
            } catch (e) {}
        }

        // #endregion

        // #region [khởi chạy server]

        if (server.koa_instance) return;

        if (!server.koa_attached) {
            server.koa.use(server.requestHandle);
            server.koa_attached = true;
        }

        server.koa_instance = server.koa.listen(server.options.port);

        // #endregion
    });
