// lixi-core-nodejs/module/api_server/middleware/request_log.js
module.exports = server =>
    server.registerMiddleware({
        key: 'request_log',
        summary: 'ghi nhận log của request',
        handler: async function({ server, request, response, api }) {
            if (!server.request_log || !server.options.request_log_enable)
                return;

            // #region [lấy thông tin]

            let meta_field = 'meta';

            if (api) {
                let option = api.middlewares.find(x => x.key === 'prepare')
                    .option_value;
                meta_field = option.meta_field;
            }

            // #endregion

            server.request_log.writeReq({
                client_name: server.options.request_log_table_name,
                id: response.body[meta_field].request_uuid,
                req: request,
                merge: {
                    request_time: response.body[meta_field].request_time,
                    response_time: response.body[meta_field].response_time,
                    process_time: response.body[meta_field].execution_time,
                    http_code: response.status,
                    html_text: response.message,
                    request_header: JSON.stringify(request.headers),
                    response_header: JSON.stringify(response.headers),
                    response_data: JSON.stringify(response.body)
                }
            });
        }
    });
