// lixi-core-nodejs/module/logger/init.js
/** khởi tạo logger
 * @param {object} options thông số cấu hình
 * @returns {WinstonLogger}} logger
 */

let Joi = require('../joi/_index');
let Winston = require('winston');
require('winston-daily-rotate-file');

// #region [schema]

let schema_option = Joi.object({
    service: Joi.string().description('tên của service'),
    module: Joi.string().description('tên của module'),
    error_stack: Joi.boolean()
        .default(true)
        .description('chi tiết lỗi'),
    level: Joi.string()
        .valid('error', 'warn', 'info', 'verbose', 'debug', 'silly')
        .default('info')
        .description('cấp độ log'),
});
schema_option = schema_option.default(schema_option.validate({}).value);

// #endregion

module.exports = expo =>
    (expo.init = function(options = {}) {
        // #region [ghi nhận option]

        options = Joi.validateValue({ options }, schema_option);

        // #endregion

        // #region [chuẩn bị các meta mặc định]

        defaultMeta = {};
        if (options.service) defaultMeta.service = options.service;
        if (options.module) defaultMeta.module = options.module;

        // #endregion

        // #region [khởi tạo logger]

        let logger = Winston.createLogger({
            level: options.level,
            format: Winston.format.combine(
                Winston.format.timestamp({ format: 'YYYY-MM-DD HH:mm:ss' }),
                Winston.format.errors({ stack: options.error_stack }),
                Winston.format.splat(),
                Winston.format.json()
            ),
            defaultMeta: defaultMeta
        });

        // #endregion

        // #region [console log]

        logger.add(
            new Winston.transports.Console({
                format: Winston.format.combine(
                    Winston.format.colorize(),
                    Winston.format.simple()
                )
            })
        );

        // #endregion

        return logger;
    });
