// lixi-core-nodejs/module/elasticsearch/query/get.js
/** thực hiện truy vấn get */

module.exports = ({ elastic, query }) =>
    (query.get = async function() {
        let result;

        if (!!query._body.id) {
            try {
                result = (
                    await elastic.client.get({
                        index: query._body.index,
                        type: query._body.type,
                        id: query._body.id
                    })
                ).body;
            } catch (e) {
                if (e.meta && e.meta.statusCode === 404) result = null;
                else throw e;
            }
        } else {
            result = await elastic.client.search({
                index: query._body.index,
                body: {
                    query: {
                        bool: {
                            must: query._body.filter_must || [],
                            should: query._body.filter_should || [],
                            must_not: query._body.filter_must_not || []
                        }
                    },
                    from: query._body.from || 0,
                    size: 1,
                    sort: query._body.sort || []
                }
            });
            result = result.body.hits.hits[0];
        }

        if (!result) return null;
        return result._source;
    });
