// lixi-core-nodejs/module/elasticsearch/query/search.js
/** thực hiện truy vấn search */

module.exports = ({ elastic, query }) =>
    (query.search = async function() {
        let data = await elastic.client.search({
            index: query._body.index,
            body: {
                query: {
                    bool: {
                        must: query._body.filter_must || [],
                        should: query._body.filter_should || [],
                        must_not: query._body.filter_must_not || []
                    }
                },
                search_after: query._body.search_after || undefined,
                from: query._body.from || 0,
                size: query._body.size || 10,
                sort: query._body.sort || []
            }
        });

        let total;
        if (!data.body.hits.total) total = 0;
        else {
            if (typeof data.body.hits.total === 'number') {
                total = data.body.hits.total;
            } else if (data.body.hits.total.value) {
                total = data.body.hits.total.value;
            } else total = 0;
        }

        let items = [];
        for (let hit of data.body.hits.hits) {
            items.push(hit._source);
        }

        return {
            total: total,
            items: items
        };
    });
